[CSL-1703] Fix space leak in readback buffer

avieth opened the pull request:
```
{-# LANGUAGE OverloadedStrings #-}

import System.Wlog

main = do
  parseLoggerConfig "log-config.yaml" >>= setupLogging
  usingLoggerName "node" logForever

logForever :: LoggerNameBox IO ()
logForever = do
    logDebug "debug"
    logForever
```

```
# log-config.yaml
severity: Info
node:
    severity: Debug
```

This program would leak memory unless the read back buffer memory queue is regularly forced by some other means. With this patch it won't leak past the bounds given (`2 * 1024 * 1024` currently). Should mention as well that the limit is actually on the number of characters in the memory queue, rather than the number of bytes, so docs should be updated.

avieth committed:
fix space leak in readback buffer

chshersh commented:
@volhovM Could you please review and approve as well?

adinapoli-iohk commented:
@avieth I haven't profiled anything in that area, but wouldn't [this line](https://github.com/serokell/log-warper/pull/17/files#diff-7b86c07e3eb88ac12fab0a935cd17e4bR58) suffer a similar malady?

avieth commented:
> @avieth I haven't profiled anything in that area, but wouldn't this line suffer a similar malady?

As we agreed, it's not so serious: the user of `readBack` will probably force the list right away, and it's not putting un-forced data into a mutable cell. Still, I'll be defensive and make it strict.

avieth committed:
readBack forces returned value

This is defensive: prevents an easy leak in case the user of readBack
doesn't force the value.

dcoutts reviewed (approved):
LGTM.
